pomegranate/gmm.pyx

```diff
@@ -138,48 +143,99 @@ cdef class GeneralMixtureModel(BayesModel):
 			model is improving per iteration.
 			Default is 1e8.
 
+		min_iterations : int, optional
+			The minimum number of iterations to run EM for. Default is 0.
+
 		pseudocount : double, optional, positive
-            A pseudocount to add to the emission of each distribution. This
-            effectively smoothes the states to prevent 0. probability symbols
-            if they don't happen to occur in the data. Only effects mixture
-            models defined over discrete distributions. Default is 0.
+			A pseudocount to add to the emission of each distribution. This
+			effectively smoothes the states to prevent 0. probability symbols
+			if they don't happen to occur in the data. Only effects mixture
+			models defined over discrete distributions. Default is 0.
 
 		verbose : bool, optional
 			Whether or not to print out improvement information over
 			iterations.
 			Default is False.
 
+		callbacks: array-like or None, optional
+			A list of Callback objects that are executed when training is
+			started and finished, as well as on both the start and finish of
+			each iteration. Defaults to None.
+
+		return_history: bool, optional
+			Signifies whether the history object is returned. Defaults to False.
+
 		Returns
 		-------
 		improvement : double
 			The total improvement in log probability P(D|M)
+
+		history : a History object, optional
+			A History object that stores the whole history of
+			training. Does not get returned unless return_history
+			is True.
+
 		"""
+		history = History()
+		callbacks = callbacks or [history]
+		callbacks.append(ParamsTerminator())
+		params = {
+		    'min_iterations': min_iterations,
+		    'max_iterations': max_iterations,
+		    'stop_threshold': stop_threshold
+		}
+		callbacks = CallbackList(callbacks=callbacks, params=params,
+		                         model=self)
+		callbacks.on_train_begin()
+		self.stop_training = False
 
 		initial_log_probability_sum = NEGINF
 		iteration, improvement = 0, INF
 
-		while improvement > stop_threshold and iteration < max_iterations + 1:
+		training_start_time = time.time()
+		while True:
+			callbacks.on_iteration_begin(iteration)
+			if self.stop_training:
```

Comment:
@jmschrei I put it there in order for the new code to be as close to the previous implementation as possible. But I agree that the better place for this may be at the end of the loop.